File: companion/src/eis_companion/site.py
"""
============================================================================
Eye in the Sky -- COMPANION site-model loader (docs/SITE_CONTRACT.md)
----------------------------------------------------------------------------
Loads the plant site model (home, perimeter geofence, no-fly zones, altitude
band, staging points) from the single site JSON file that the whole system
shares. Nothing outside ``site/`` may hardcode plant geometry -- consumers get
it from here as plain dataclasses.

File selection (mirrors docs/SITE_CONTRACT.md):
  * ``load_site(path)`` takes an EXPLICIT path -- the orchestrator passes the
    configured value.
  * ``resolve_site_path()`` resolves the conventional location: explicit arg,
    then the ``EIS_SITE_FILE`` env var, then the default ``site/site.json`` --
    relative paths are anchored at the repo root.
  * ``load_site_from_env()`` composes the two for callers that just want the
    conventional file.

Schema semantics (binding, see docs/SITE_CONTRACT.md):
  * Coordinates are WGS84, ordered ``[lat, lon]`` (never lon-first).
  * Polygons are open rings (closing edge implicit); either winding is fine,
    so this loader preserves vertex order verbatim.
  * ``alt_band_m`` is metres AGL relative to home; it may only tighten, never
    relax, the hard limits asserted in ``config.py`` (enforced there, not here).
  * ``nfz[].ceiling_m``: flight inside the polygon at/below the ceiling is
    forbidden; overflight above it is permitted.
  * ``perimeter`` is the outer geofence -- uploaded to ArduPilot's polygon
    fence on connect (``Vehicle.upload_geofence``) and checked independently
    ground-side.

This module deliberately lives OUTSIDE ``control/``: it does file I/O.
Pure-logic control code must receive site data as plain arguments, never read
the file itself. Pure stdlib only (json / os / pathlib / dataclasses).

Validation errors raise ``ValueError`` with a message naming the offending
field. A missing file raises the natural ``FileNotFoundError``.
============================================================================
"""
from __future__ import annotations

import json
import math
import os
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, List, Optional, Tuple

# Env var + default location per docs/SITE_CONTRACT.md ("File selection").
SITE_FILE_ENV = "EIS_SITE_FILE"
DEFAULT_SITE_REL = "site/site.json"

# Ground-truth labels a staging point may carry (SITE_CONTRACT schema).
TRUTH_LABELS = ("vehicle", "breach", "structure", "false_alarm")

LatLon = Tuple[float, float]


# --------------------------------------------------------------------------
# Parsed site model -- simple dataclasses, plain data only
# --------------------------------------------------------------------------
@dataclass
class Home:
    """The home/launch point. ``alt_m`` is terrain elevation in metres AMSL."""
    lat: float
    lon: float
    alt_m: float


@dataclass
class NoFlyZone:
    """A polygon NFZ: flight inside at or below ``ceiling_m`` AGL is forbidden."""
    name: str
    polygon: List[LatLon]        # open ring, [lat, lon] order preserved
    ceiling_m: float


@dataclass
class AltBand:
    """Permitted flight band in metres AGL relative to home."""
    min_m: float
    max_m: float


@dataclass
class StagingPoint:
    """A pre-surveyed observation point with its still image + truth label."""
    id: str
    lat: float
    lon: float
    image: str                   # repo-root-relative path to the still
    truth: str                   # one of TRUTH_LABELS


@dataclass
class Site:
    """The full parsed site model (docs/SITE_CONTRACT.md schema)."""
    home: Home
    perimeter: List[LatLon]      # outer geofence, open ring of (lat, lon)
    nfz: List[NoFlyZone] = field(default_factory=list)
    alt_band: AltBand = field(default_factory=lambda: AltBand(0.0, 0.0))
    staging: List[StagingPoint] = field(default_factory=list)
    source_path: Optional[str] = None   # where this model was loaded from


# --------------------------------------------------------------------------
# Path resolution (explicit arg > EIS_SITE_FILE > site/site.json @ repo root)
# --------------------------------------------------------------------------
def _repo_root() -> Path:
    """The monorepo root (…/companion/src/eis_companion -> up 3)."""
    return Path(__file__).resolve().parents[3]


def resolve_site_path(explicit: Optional[str] = None) -> Path:
    """Resolve which site JSON to load. Does not require the file to exist.

    Order: ``explicit`` arg, then the ``EIS_SITE_FILE`` env var, then the
    default ``site/site.json``. Relative candidates that don't resolve from
    the current working directory are anchored at the repo root, per
    docs/SITE_CONTRACT.md ("path … relative to the repo root").
    """
    candidate = explicit or os.environ.get(SITE_FILE_ENV) or DEFAULT_SITE_REL
    p = Path(candidate)
    if p.is_absolute() or p.is_file():
        return p
    return _repo_root() / candidate


# --------------------------------------------------------------------------
# Loading + validation
# --------------------------------------------------------------------------
def load_site(path: "str | os.PathLike[str]") -> Site:
    """Parse + validate the site JSON at ``path`` into a :class:`Site`.

    The path is EXPLICIT -- the caller passes its configured value (use
    :func:`resolve_site_path` / :func:`load_site_from_env` for the
    conventional env-var lookup). Raises ``ValueError`` on schema violations
    and ``FileNotFoundError`` when the file is missing.
    """
    p = Path(path)
    with p.open("r", encoding="utf-8") as f:
        try:
            raw = json.load(f)
        except json.JSONDecodeError as exc:
            raise ValueError(f"site: {p} is not valid JSON: {exc}") from exc
    if not isinstance(raw, dict):
        raise ValueError(f"site: {p} must contain a JSON object at the top level")

    home = _parse_home(raw.get("home"))
    perimeter = _parse_polygon(raw.get("perimeter"), "perimeter")
    nfz = _parse_nfz(raw.get("nfz"))
    alt_band = _parse_alt_band(raw.get("alt_band_m"))
    staging = _parse_staging(raw.get("staging"))

    return Site(
        home=home,
        perimeter=perimeter,
        nfz=nfz,
        alt_band=alt_band,
        staging=staging,
        source_path=str(p),
    )


def load_site_from_env(explicit: Optional[str] = None) -> Site:
    """Load the site model from the conventional location.

    ``explicit`` (if given) wins, else ``EIS_SITE_FILE``, else
    ``site/site.json`` relative to the repo root.
    """
    return load_site(resolve_site_path(explicit))


# --------------------------------------------------------------------------
# Field parsers -- each raises ValueError naming the offending field
# --------------------------------------------------------------------------
def _num(value: Any, ctx: str) -> float:
    """A finite float, or ValueError."""
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        raise ValueError(f"site: {ctx} must be a number, got {value!r}")
    f = float(value)
    if not math.isfinite(f):
        raise ValueError(f"site: {ctx} must be finite, got {value!r}")
    return f


def _latlon(lat: Any, lon: Any, ctx: str) -> LatLon:
    """A valid (lat, lon) pair, or ValueError. Order is [lat, lon] always."""
    lat_f = _num(lat, f"{ctx}.lat")
    lon_f = _num(lon, f"{ctx}.lon")
    if not -90.0 <= lat_f <= 90.0:
        raise ValueError(f"site: {ctx}.lat out of range [-90, 90]: {lat_f}")
    if not -180.0 <= lon_f <= 180.0:
        raise ValueError(f"site: {ctx}.lon out of range [-180, 180]: {lon_f}")
    return (lat_f, lon_f)


def _parse_home(raw: Any) -> Home:
    if not isinstance(raw, dict):
        raise ValueError("site: 'home' object is required")
    lat, lon = _latlon(raw.get("lat"), raw.get("lon"), "home")
    return Home(lat=lat, lon=lon, alt_m=_num(raw.get("alt_m"), "home.alt_m"))


def _parse_polygon(raw: Any, ctx: str, *, min_vertices: int = 3) -> List[LatLon]:
    """An open ring of [lat, lon] pairs (vertex order preserved verbatim)."""
    if not isinstance(raw, list):
        raise ValueError(f"site: '{ctx}' must be a list of [lat, lon] pairs")
    if len(raw) < min_vertices:
        raise ValueError(
            f"site: '{ctx}' needs at least {min_vertices} vertices, got {len(raw)}"
        )
    ring: List[LatLon] = []
    for i, pair in enumerate(raw):
        if not isinstance(pair, (list, tuple)) or len(pair) != 2:
            raise ValueError(f"site: {ctx}[{i}] must be a [lat, lon] pair")
        ring.append(_latlon(pair[0], pair[1], f"{ctx}[{i}]"))
    return ring


def _parse_nfz(raw: Any) -> List[NoFlyZone]:
    if raw is None:
        return []
    if not isinstance(raw, list):
        raise ValueError("site: 'nfz' must be a list")
    zones: List[NoFlyZone] = []
    for i, z in enumerate(raw):
        if not isinstance(z, dict):
            raise ValueError(f"site: nfz[{i}] must be an object")
        name = z.get("name")
        if not isinstance(name, str) or not name:
            raise ValueError(f"site: nfz[{i}].name must be a non-empty string")
        ceiling = _num(z.get("ceiling_m"), f"nfz[{i}].ceiling_m")
        if ceiling < 0.0:
            raise ValueError(f"site: nfz[{i}].ceiling_m must be >= 0, got {ceiling}")
        polygon = _parse_polygon(z.get("polygon"), f"nfz[{i}].polygon")
        zones.append(NoFlyZone(name=name, polygon=polygon, ceiling_m=ceiling))
    return zones


def _parse_alt_band(raw: Any) -> AltBand:
    if not isinstance(raw, dict):
        raise ValueError("site: 'alt_band_m' object is required")
    lo = _num(raw.get("min"), "alt_band_m.min")
    hi = _num(raw.get("max"), "alt_band_m.max")
    if lo < 0.0:
        raise ValueError(f"site: alt_band_m.min must be >= 0, got {lo}")
    if hi < lo:
        raise ValueError(f"site: alt_band_m.max ({hi}) < min ({lo})")
    return AltBand(min_m=lo, max_m=hi)


def _parse_staging(raw: Any) -> List[StagingPoint]:
    if raw is None:
        return []
    if not isinstance(raw, list):
        raise ValueError("site: 'staging' must be a list")
    points: List[StagingPoint] = []
    seen_ids: set = set()
    for i, s in enumerate(raw):
        if not isinstance(s, dict):
            raise ValueError(f"site: staging[{i}] must be an object")
        sid = s.get("id")
        if not isinstance(sid, str) or not sid:
            raise ValueError(f"site: staging[{i}].id must be a non-empty string")
        if sid in seen_ids:
            raise ValueError(f"site: staging[{i}].id duplicates {sid!r}")
        seen_ids.add(sid)
        lat, lon = _latlon(s.get("lat"), s.get("lon"), f"staging[{i}]")
        image = s.get("image")
        if not isinstance(image, str) or not image:
            raise ValueError(f"site: staging[{i}].image must be a non-empty path")
        truth = s.get("truth")
        if truth not in TRUTH_LABELS:
            raise ValueError(
                f"site: staging[{i}].truth must be one of {TRUTH_LABELS}, "
                f"got {truth!r}"
            )
        points.append(StagingPoint(id=sid, lat=lat, lon=lon, image=image, truth=truth))
    return points


__all__ = [
    "SITE_FILE_ENV",
    "DEFAULT_SITE_REL",
    "TRUTH_LABELS",
    "Home",
    "NoFlyZone",
    "AltBand",
    "StagingPoint",
    "Site",
    "resolve_site_path",
    "load_site",
    "load_site_from_env",
]

File: ground/planner/src/site.ts
/* ============================================================================
 * eis-planner/site — site-model loading, validation, and geometry helpers.
 *
 * The site model is consumed ONLY via the site JSON file described in
 * docs/SITE_CONTRACT.md. Nothing in this package hardcodes plant geometry —
 * every coordinate is derived from the loaded site data.
 *
 * Units and conventions (see docs/SITE_CONTRACT.md):
 *  - Coordinates are WGS84, ordered [lat, lon] in the JSON (lat first).
 *  - Polygons are OPEN rings (closing edge implicit); either winding is valid.
 *  - `alt_band_m` and NFZ `ceiling_m` are meters AGL relative to home.
 *  - `perimeter` is the outer geofence.
 *  - NFZ semantics: flight inside the polygon AT OR BELOW `ceiling_m` AGL is
 *    forbidden; overflight strictly above the ceiling is permitted.
 *
 * Geometry helpers all work on {lat, lon} degree pairs. Internally they use a
 * local equirectangular projection (meters east/north of a reference point) —
 * accurate to well under 0.1% at plant scale (< a few km), which is far inside
 * the safety margins used by the verifier.
 * ========================================================================== */

import * as fs from 'fs';
import * as path from 'path';

/* ---------------------------------------------------------------------------
 * Types
 * ------------------------------------------------------------------------- */

/** A WGS84 point in decimal degrees. */
export interface LatLon {
  lat: number;
  lon: number;
}

/** Wire form used by the site JSON: [lat, lon] (lat FIRST — never lon-first). */
export type LatLonPair = [number, number];

export interface SiteNfz {
  name: string;
  /** Open ring, any winding. */
  polygon: LatLon[];
  /** Forbidden at or below this altitude, meters AGL. */
  ceilingM: number;
}

export interface SiteStagingPoint {
  id: string;
  lat: number;
  lon: number;
  /** Repo-root-relative path to the still image. */
  image: string;
  truth: 'vehicle' | 'breach' | 'structure' | 'false_alarm';
}

export interface SiteModel {
  home: { lat: number; lon: number; altM: number };
  /** Outer geofence — open ring, any winding. */
  perimeter: LatLon[];
  nfz: SiteNfz[];
  /** Permitted flight band, meters AGL relative to home. */
  altBandM: { min: number; max: number };
  staging: SiteStagingPoint[];
}

/* ---------------------------------------------------------------------------
 * File selection + loading
 * ------------------------------------------------------------------------- */

/** Default site file, relative to the repo root (docs/SITE_CONTRACT.md). */
export const DEFAULT_SITE_FILE = 'site/site.json';

/**
 * Resolve the site JSON path: `EIS_SITE_FILE` env var (repo-root-relative or
 * absolute) falling back to `site/site.json`. `repoRoot` defaults to the
 * current working directory. Pure path math — does NOT check existence.
 */
export function resolveSiteFile(repoRoot: string = process.cwd()): string {
  const rel = process.env.EIS_SITE_FILE && process.env.EIS_SITE_FILE.trim() !== ''
    ? process.env.EIS_SITE_FILE
    : DEFAULT_SITE_FILE;
  return path.isAbsolute(rel) ? rel : path.resolve(repoRoot, rel);
}

/** Load + validate the site JSON at `filePath`. Throws on any invalid input. */
export function loadSite(filePath: string): SiteModel {
  let raw: string;
  try {
    raw = fs.readFileSync(filePath, 'utf8');
  } catch (err) {
    throw new Error(`site file not readable: ${filePath} (${(err as Error).message})`);
  }
  let data: unknown;
  try {
    data = JSON.parse(raw);
  } catch (err) {
    throw new Error(`site file is not valid JSON: ${filePath} (${(err as Error).message})`);
  }
  return validateSite(data);
}

/** Convenience: load the EIS_SITE_FILE-selected site (default site/site.json). */
export function loadSiteFromEnv(repoRoot: string = process.cwd()): SiteModel {
  return loadSite(resolveSiteFile(repoRoot));
}

/* ---------------------------------------------------------------------------
 * Validation
 * ------------------------------------------------------------------------- */

const TRUTH_VALUES = new Set(['vehicle', 'breach', 'structure', 'false_alarm']);

function isFiniteNumber(v: unknown): v is number {
  return typeof v === 'number' && Number.isFinite(v);
}

function assertLat(v: unknown, ctx: string): number {
  if (!isFiniteNumber(v) || v < -90 || v > 90) {
    throw new Error(`invalid site JSON: ${ctx} must be a latitude in [-90, 90], got ${JSON.stringify(v)}`);
  }
  return v;
}

function assertLon(v: unknown, ctx: string): number {
  if (!isFiniteNumber(v) || v < -180 || v > 180) {
    throw new Error(`invalid site JSON: ${ctx} must be a longitude in [-180, 180], got ${JSON.stringify(v)}`);
  }
  return v;
}

function parseRing(v: unknown, ctx: string): LatLon[] {
  if (!Array.isArray(v) || v.length < 3) {
    throw new Error(`invalid site JSON: ${ctx} must be an array of >= 3 [lat, lon] pairs`);
  }
  return v.map((pair, i) => {
    if (!Array.isArray(pair) || pair.length !== 2) {
      throw new Error(`invalid site JSON: ${ctx}[${i}] must be a [lat, lon] pair`);
    }
    return {
      lat: assertLat(pair[0], `${ctx}[${i}][0]`),
      lon: assertLon(pair[1], `${ctx}[${i}][1]`),
    };
  });
}

/**
 * Validate an already-parsed site object into a typed SiteModel.
 * Throws a descriptive Error on any schema/semantic violation.
 */
export function validateSite(data: unknown): SiteModel {
  if (typeof data !== 'object' || data === null) {
    throw new Error('invalid site JSON: root must be an object');
  }
  const d = data as Record<string, unknown>;

  const home = d.home as Record<string, unknown> | undefined;
  if (typeof home !== 'object' || home === null) {
    throw new Error('invalid site JSON: missing "home" object');
  }
  const homeLat = assertLat(home.lat, 'home.lat');
  const homeLon = assertLon(home.lon, 'home.lon');
  if (!isFiniteNumber(home.alt_m)) {
    throw new Error('invalid site JSON: home.alt_m must be a finite number (meters AMSL)');
  }

  const perimeter = parseRing(d.perimeter, 'perimeter');

  const nfzRaw = d.nfz;
  if (!Array.isArray(nfzRaw)) {
    throw new Error('invalid site JSON: "nfz" must be an array (may be empty)');
  }
  const nfz: SiteNfz[] = nfzRaw.map((z, i) => {
    const zz = z as Record<string, unknown>;
    if (typeof zz !== 'object' || zz === null || typeof zz.name !== 'string') {
      throw new Error(`invalid site JSON: nfz[${i}] must be an object with a string "name"`);
    }
    if (!isFiniteNumber(zz.ceiling_m) || (zz.ceiling_m as number) < 0) {
      throw new Error(`invalid site JSON: nfz[${i}].ceiling_m must be a number >= 0 (meters AGL)`);
    }
    return {
      name: zz.name,
      polygon: parseRing(zz.polygon, `nfz[${i}].polygon`),
      ceilingM: zz.ceiling_m as number,
    };
  });

  const band = d.alt_band_m as Record<string, unknown> | undefined;
  if (typeof band !== 'object' || band === null ||
      !isFiniteNumber(band.min) || !isFiniteNumber(band.max)) {
    throw new Error('invalid site JSON: "alt_band_m" must be { min, max } in meters AGL');
  }
  if ((band.min as number) < 0 || (band.min as number) >= (band.max as number)) {
    throw new Error(`invalid site JSON: alt_band_m requires 0 <= min < max, got min=${band.min} max=${band.max}`);
  }

  const stagingRaw = d.staging;
  if (!Array.isArray(stagingRaw)) {
    throw new Error('invalid site JSON: "staging" must be an array (may be empty)');
  }
  const staging: SiteStagingPoint[] = stagingRaw.map((s, i) => {
    const ss = s as Record<string, unknown>;
    if (typeof ss !== 'object' || ss === null || typeof ss.id !== 'string' ||
        typeof ss.image !== 'string' || typeof ss.truth !== 'string' ||
        !TRUTH_VALUES.has(ss.truth)) {
      throw new Error(
        `invalid site JSON: staging[${i}] must have string id/image and truth in ` +
        `vehicle|breach|structure|false_alarm`,
      );
    }
    return {
      id: ss.id,
      lat: assertLat(ss.lat, `staging[${i}].lat`),
      lon: assertLon(ss.lon, `staging[${i}].lon`),
      image: ss.image,
      truth: ss.truth as SiteStagingPoint['truth'],
    };
  });

  return {
    home: { lat: homeLat, lon: homeLon, altM: home.alt_m as number },
    perimeter,
    nfz,
    altBandM: { min: band.min as number, max: band.max as number },
    staging,
  };
}

/* ---------------------------------------------------------------------------
 * Geometry helpers
 *
 * All distances are METERS; all coordinates are DEGREES ({lat, lon}).
 * ------------------------------------------------------------------------- */

/** Mean earth radius, meters (WGS84 mean radius). */
export const EARTH_RADIUS_M = 6371000;

/** Meters per degree of latitude (near-constant over the globe). */
export const M_PER_DEG_LAT = 111320;

const DEG = Math.PI / 180;

/**
 * Great-circle distance between two WGS84 points, in METERS (haversine
 * formula on a sphere of radius EARTH_RADIUS_M).
 */
export function haversineMeters(a: LatLon, b: LatLon): number {
  const dLat = (b.lat - a.lat) * DEG;
  const dLon = (b.lon - a.lon) * DEG;
  const sinLat = Math.sin(dLat / 2);
  const sinLon = Math.sin(dLon / 2);
  const h = sinLat * sinLat +
    Math.cos(a.lat * DEG) * Math.cos(b.lat * DEG) * sinLon * sinLon;
  return 2 * EARTH_RADIUS_M * Math.asin(Math.min(1, Math.sqrt(h)));
}

interface XY { x: number; y: number; } // meters east (x) / north (y) of origin

/** Project a point to local meters (equirectangular, anchored at `origin`). */
function toLocal(p: LatLon, origin: LatLon): XY {
  return {
    x: (p.lon - origin.lon) * M_PER_DEG_LAT * Math.cos(origin.lat * DEG),
    y: (p.lat - origin.lat) * M_PER_DEG_LAT,
  };
}

/** Inverse of toLocal. */
function fromLocal(p: XY, origin: LatLon): LatLon {
  return {
    lat: origin.lat + p.y / M_PER_DEG_LAT,
    lon: origin.lon + p.x / (M_PER_DEG_LAT * Math.cos(origin.lat * DEG)),
  };
}

/**
 * Point-in-polygon (ray casting) for an OPEN ring in degree space.
 * Winding-agnostic. Points exactly on an edge are implementation-defined
 * (treated as inside/outside depending on float rounding) — the verifier
 * always applies a margin, so exact-boundary points never matter in practice.
 */
export function pointInPolygon(p: LatLon, polygon: LatLon[]): boolean {
  const origin = polygon[0];
  const pt = toLocal(p, origin);
  const ring = polygon.map((v) => toLocal(v, origin));
  let inside = false;
  for (let i = 0, j = ring.length - 1; i < ring.length; j = i++) {
    const a = ring[i];
    const b = ring[j];
    const crosses = (a.y > pt.y) !== (b.y > pt.y);
    if (crosses) {
      const xAtY = a.x + ((pt.y - a.y) / (b.y - a.y)) * (b.x - a.x);
      if (pt.x < xAtY) inside = !inside;
    }
  }
  return inside;
}

function orient(a: XY, b: XY, c: XY): number {
  return (b.x - a.x) * (c.y - a.y) - (b.y - a.y) * (c.x - a.x);
}

function onSegment(a: XY, b: XY, p: XY): boolean {
  return Math.min(a.x, b.x) - 1e-9 <= p.x && p.x <= Math.max(a.x, b.x) + 1e-9 &&
         Math.min(a.y, b.y) - 1e-9 <= p.y && p.y <= Math.max(a.y, b.y) + 1e-9;
}

function segmentsIntersect(p1: XY, p2: XY, p3: XY, p4: XY): boolean {
  const d1 = orient(p3, p4, p1);
  const d2 = orient(p3, p4, p2);
  const d3 = orient(p1, p2, p3);
  const d4 = orient(p1, p2, p4);
  if (((d1 > 0 && d2 < 0) || (d1 < 0 && d2 > 0)) &&
      ((d3 > 0 && d4 < 0) || (d3 < 0 && d4 > 0))) {
    return true;
  }
  if (d1 === 0 && onSegment(p3, p4, p1)) return true;
  if (d2 === 0 && onSegment(p3, p4, p2)) return true;
  if (d3 === 0 && onSegment(p1, p2, p3)) return true;
  if (d4 === 0 && onSegment(p1, p2, p4)) return true;
  return false;
}

/**
 * True when the segment a→b intersects ANY edge of the (open-ring) polygon,
 * including the implicit closing edge. A segment lying entirely inside or
 * entirely outside the polygon does NOT intersect its boundary.
 */
export function segmentIntersectsPolygon(a: LatLon, b: LatLon, polygon: LatLon[]): boolean {
  const origin = polygon[0];
  const pa = toLocal(a, origin);
  const pb = toLocal(b, origin);
  const ring = polygon.map((v) => toLocal(v, origin));
  for (let i = 0; i < ring.length; i++) {
    const e1 = ring[i];
    const e2 = ring[(i + 1) % ring.length];
    if (segmentsIntersect(pa, pb, e1, e2)) return true;
  }
  return false;
}

/**
 * True when the whole segment a→b stays strictly inside a SIMPLE polygon:
 * both endpoints inside AND no boundary crossing. (For a simple polygon a
 * segment can only leave the interior by crossing the boundary.)
 */
export function segmentStaysInsidePolygon(a: LatLon, b: LatLon, polygon: LatLon[]): boolean {
  return pointInPolygon(a, polygon) &&
         pointInPolygon(b, polygon) &&
         !segmentIntersectsPolygon(a, b, polygon);
}

/**
 * True when the segment a→b enters the polygon at all: either endpoint
 * inside, or the segment crosses the boundary (pass-through counts).
 */
export function segmentEntersPolygon(a: LatLon, b: LatLon, polygon: LatLon[]): boolean {
  return pointInPolygon(a, polygon) ||
         pointInPolygon(b, polygon) ||
         segmentIntersectsPolygon(a, b, polygon);
}

function nearestPointOnSegmentXY(p: XY, a: XY, b: XY): XY {
  const abx = b.x - a.x;
  const aby = b.y - a.y;
  const len2 = abx * abx + aby * aby;
  if (len2 === 0) return { x: a.x, y: a.y };
  let t = ((p.x - a.x) * abx + (p.y - a.y) * aby) / len2;
  t = Math.max(0, Math.min(1, t));
  return { x: a.x + t * abx, y: a.y + t * aby };
}

/** Closest point on the polygon's boundary (including the closing edge). */
export function nearestPointOnPolygonBoundary(p: LatLon, polygon: LatLon[]): LatLon {
  const origin = polygon[0];
  const pt = toLocal(p, origin);
  const ring = polygon.map((v) => toLocal(v, origin));
  let best: XY = ring[0];
  let bestD2 = Infinity;
  for (let i = 0; i < ring.length; i++) {
    const cand = nearestPointOnSegmentXY(pt, ring[i], ring[(i + 1) % ring.length]);
    const dx = cand.x - pt.x;
    const dy = cand.y - pt.y;
    const d2 = dx * dx + dy * dy;
    if (d2 < bestD2) {
      bestD2 = d2;
      best = cand;
    }
  }
  return fromLocal(best, origin);
}

/**
 * Move `p` to the nearest polygon-boundary point and then continue `marginM`
 * meters FURTHER along the same direction (p → boundary → beyond).
 *
 *  - p INSIDE the polygon  → result is OUTSIDE it by ~marginM (NFZ push-out).
 *  - p OUTSIDE the polygon → result is INSIDE it by ~marginM (perimeter pull-in).
 *
 * Degenerate case (p exactly on the boundary / zero distance): the direction
 * is taken from the polygon's vertex centroid through p instead, which points
 * outward for convex polygons. This is best-effort for pathological input.
 */
export function movePointAcrossBoundary(p: LatLon, polygon: LatLon[], marginM: number): LatLon {
  const origin = polygon[0];
  const pt = toLocal(p, origin);
  const boundary = toLocal(nearestPointOnPolygonBoundary(p, polygon), origin);
  let dx = boundary.x - pt.x;
  let dy = boundary.y - pt.y;
  let len = Math.hypot(dx, dy);
  if (len < 1e-9) {
    // p is on the boundary: fall back to centroid→p direction.
    const ring = polygon.map((v) => toLocal(v, origin));
    const cx = ring.reduce((s, v) => s + v.x, 0) / ring.length;
    const cy = ring.reduce((s, v) => s + v.y, 0) / ring.length;
    dx = pt.x - cx;
    dy = pt.y - cy;
    len = Math.hypot(dx, dy);
    if (len < 1e-9) {
      dx = 1; dy = 0; len = 1; // fully degenerate: arbitrary but deterministic east
    }
  }
  const scale = marginM / len;
  return fromLocal({ x: boundary.x + dx * scale, y: boundary.y + dy * scale }, origin);
}

/** Vertex centroid of a polygon (adequate for target placement, not area math). */
export function polygonCentroid(polygon: LatLon[]): LatLon {
  const lat = polygon.reduce((s, v) => s + v.lat, 0) / polygon.length;
  const lon = polygon.reduce((s, v) => s + v.lon, 0) / polygon.length;
  return { lat, lon };
}
